plookup/src/table/hash_tables/tables.rs

```diff
@@ -0,0 +1,207 @@
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+///! In circuit hashing can be conducted using table lookups from the
+///! tables defined in this file.
+/// This currently assumes that T_S = 4. It is possible that in future T_S > 4,
+/// in which case this file will have to be adjusted. T_S is the arity of the
+/// vectors.
+use crate::table::hash_tables::constants::{N, S, T_S, V};
+use dusk_plonk::fft::Polynomial;
+use dusk_plonk::prelude::BlsScalar;
+use std::convert::TryInto;
+
+/// A HashTable consists of three different segments, each of arity
+/// vectors containing Bls Scalars. These values are precomputed and
+/// stored. These are the retrieved in the lookup functions and placed
+/// into seperate structs for the proof.
+/// A vector x in (F_p)^t goes through r rounds of a round function R.
+/// The result is another vector y in (F_p)^t.
+#[derive(Debug)]
+pub struct HashTable {
+    pub first_rows: Vec<[BlsScalar; 4]>,
+    pub middle_rows: Vec<[BlsScalar; 4]>,
+    pub end_rows: Vec<[BlsScalar; 4]>,
+}
+
+impl HashTable {
+    /// Create a new hash table and makes
+    /// empty vectors for each field.
+    pub fn new() -> Self {
+        Self {
+            first_rows: vec![],
+            middle_rows: vec![],
+            end_rows: vec![],
+        }
+    }
+
+    // The whole lookup table will be constructed in 3 parts: the first rows where the third
+    // entry is derived from the function F, i.e. the rows are of the form (_, _, F(i), ...).
+    // The middle rows are where the first entries are between V+1 and s_i for some i.
+    // The binary rows are at the bottom of the table, and they enumerate all binary possibilities
+    // on T_S bits.
+    fn f_rows(&mut self, f: Polynomial) {
+        for i in 0..(V + 1) {
+            let eval = f.evaluate(&BlsScalar::from(i as u64));
+            let row = [
+                BlsScalar::from(i as u64),
+                BlsScalar::zero(),
+                eval,
+                -BlsScalar::one(),
+            ];
+
+            self.first_rows.push(row);
+        }
+    }
+
+    // This function fills in the middle section of the hash table
+    // where the entry is defined as being between V+1 and s_i
+    // for a chosen i. The i here depends on the intialisation
+    // of the first rows.
+    fn m_rows(&mut self) {
+        for i in 0..N {
+            let distance = S[i as usize] - V as u64;
+
+            for j in 1..(distance + 1) {
+                let row = [
+                    BlsScalar::from(V as u64 + j),
+                    BlsScalar::from(i as u64 + 1),
+                    BlsScalar::from(V as u64 + j),
+                    -BlsScalar::one(),
+                ];
+
+                self.middle_rows.push(row);
+            }
+        }
+    }
+
+    // A function that creates all binary values of word of length T_S.
+    // For a width T_S, these are the binary possibilites for numbers
+    // between 0 and 15.
+    fn binary_end_rows(&mut self) {
```

Comment:
Is this fn needed? It only pushes empty rows so this could be done directly in the generation of `Self`.